rules-tests/Php80/Rector/FunctionLike/UnionTypesRector/Fixture/union_self_this_final_class.php.inc

```diff
@@ -12,19 +12,3 @@ final class SkipUnionSelfThisFinalClass
         return $this;
     }
 }
-
-?>
------
-<?php
-
-namespace Rector\Tests\Php80\Rector\FunctionLike\UnionTypesRector\Fixture;
-
-final class SkipUnionSelfThisFinalClass
-{
-    public function withFromId(int $fromId): self
-    {
-        return $this;
-    }
-}
-
-?>
```

Comment:
This fixture is correct. In case of `final` class, the doc is exact duplicate of self.
We've removed many of these cases manually, that's why this rule handled it.

Reply:
But `self !== $this`. Correct me if I'm wrong but the proper usage and difference is:

For mutable withers, you'd use `$this`. For immutable withers, you'd use `self`.

```php
final class A
{
    public function withMute(): self
    {
        return clone $this;
    }

    /**
     * @return $this
     */
    public function withNoMute(): self
    {
        return $this;
    }
}
```

> We've removed many of these cases manually, that's why this rule handled it.

So it depends what were your cases. If you had mutable withers, those should not have been removed AFAIK.

Reply:
It depends how PHPStan handles these 2 cases, as we depend on its type system. Is it same or different from PHPStan view?

The change might be needed there.

Reply:
I'd it does not care about that https://phpstan.org/r/54c6b6d0-1979-4ee6-9a62-927bc10317c1

Reply:
But even though phpstan might not distinguish between that, the union rule should not touch it. 🤔

Reply:
Thanks for checking and I agree with you.

But in that case it must be resolved there first. Then we can use it here.

Reply:
> But in that case it must be resolved there first. Then we can use it here.

Shouldn't this https://github.com/rectorphp/rector-src/pull/1246 be reverted then? It's not working properly, I'd say there's a bug in behaviour now.